RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterECLCC.h

```diff
@@ -0,0 +1,297 @@
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusteringEdgeVarsDeviceCollection::View pfClusteringEdgeVars) const {
+      const int lane = alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] % alpaka::warp::getSize(acc);
+
+      int32_t idx = 0;
+      if (lane == 0)
+        idx = alpaka::atomicAdd(acc, &pfClusteringVars.posL(), 1);
+      idx = alpaka::warp::shfl(acc, idx, 0);
+      while (idx < pfClusteringVars.topL()) {
+        const int v = pfClusteringVars[idx].wl_d();
+        int vstat = representative(v, pfClusteringVars);
+        for (int i = pfClusteringEdgeVars[v].pfrh_edgeIdx() + lane; i < pfClusteringEdgeVars[v + 1].pfrh_edgeIdx();
+             i += alpaka::warp::getSize(acc)) {
+          const int nli = pfClusteringEdgeVars[i].pfrh_edgeList();
+          if (v > nli) {
+            int ostat = representative(nli, pfClusteringVars);
+            bool repeat;
+            do {
+              repeat = false;
+              if (vstat != ostat) {
+                int ret;
+                if (vstat < ostat) {
+                  if ((ret = alpaka::atomicCas(acc, &pfClusteringVars[ostat].pfrh_topoId(), ostat, vstat)) != ostat) {
+                    ostat = ret;
+                    repeat = true;
+                  }
+                } else {
+                  if ((ret = alpaka::atomicCas(acc, &pfClusteringVars[vstat].pfrh_topoId(), vstat, ostat)) != vstat) {
+                    vstat = ret;
+                    repeat = true;
+                  }
+                }
+              }
+            } while (repeat);
+          }
+        }
+        if (lane == 0)
+          idx = alpaka::atomicAdd(acc, &pfClusteringVars.posL(), 1);
+        idx = alpaka::warp::shfl(acc, idx, 0);
+      }
+    }
+  };
+
+  /* process high-degree vertices at block granularity */
+  class ECLCCCompute3 {
+  public:
+    template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+    ALPAKA_FN_ACC void operator()(const TAcc& acc,
+                                  const reco::PFRecHitHostCollection::ConstView pfRecHits,
+                                  reco::PFClusteringVarsDeviceCollection::View pfClusteringVars,
+                                  reco::PFClusteringEdgeVarsDeviceCollection::View pfClusteringEdgeVars) const {
+      int& vB = alpaka::declareSharedVar<int, __COUNTER__>(acc);
+      if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0) {
+        const int idx = alpaka::atomicAdd(acc, &pfClusteringVars.posH(), -1);
+        vB = (idx > pfClusteringVars.topH()) ? pfClusteringVars[idx].wl_d() : -1;
+      }
+      alpaka::syncBlockThreads(acc);  // all threads call sync
+      while (vB >= 0) {
+        const int v = vB;
+        alpaka::syncBlockThreads(acc);  // all threads call sync
```

Comment:
do you need this synchronisation ?
there is already one right before the loop, and one at the end of each loop iteration